### PR TITLE
KT-22334 Specialized loops using range such as integer..array.size-1

### PR DESCRIPTION
**Pull request by mikaelpeltier**
- Into PrimitiveNumberRangeLiteralRangeValue modifies how bounded
value are created by checking if the high bound range can be modified
to be exclusive instead of inclusive such as the generated code will
be optimized.
- Add black box tests checking that the specialization does not fail
on any kind of arrays.
- Add a bytecode test checking that the code is correctly optimized.

Fix of https://youtrack.jetbrains.com/issue/KT-22334

### CODE REVIEW

**Comment by mikaelpeltier** [comment]
This PR was tested by:
./gradlew dist
./gradlew jvmCompilerTest

The following test failed with my PR but also without, thus it seems not related to my change.
org.jetbrains.kotlin.jvm.compiler.CompileKotlinAgainstJavaTestGenerated.testListImpl

File: compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt [hunk 1]
```diff
@@ -0,0 +1,143 @@
+// WITH_RUNTIME
+
+fun checkByteArray():Boolean {
```
Comment (jshvarts) [inline compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt]: here and in other functions below, add space before Boolean return type (`fun checkByteArray(): Boolean`)

File: compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    var start = 0
+    var sum = 0
+    for (i in (start..intArray.size-1).reversed()) {
+        sum += intArray[i]
+    }
+    if (sum != 6) return false
+    return true
+}
+
+fun box() : String {
```
Comment (jshvarts) [inline compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt]: spacing fun box(): String

File: compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt (hunk 1 above)
Comment (mikaelpeltier) [reply compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt]: Done, I reformatted the test

File: compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt (hunk 2 above)
Comment (mikaelpeltier) [reply compiler/testData/codegen/box/controlStructures/forInArray/forInArraySpecializedToUntil.kt]: Done, I reformatted the test

**Comment by mikaelpeltier** [comment]
Pushed to master by dnpetrov

**Pull request closed without merging**